### ISSUE TITLE
Invalid Focus Sequence on Android

### ISSUE THREAD

**Issue by ebarlas** [libsdl-org/SDL#8369]
I recently released an SDL-based mobile game on iOS and Android called Trippin on Tubs.

Overall, it has been completely smooth release process. SDL is an amazing foundation! So thank you, SDL contributors.

However, I noticed something odd on some Android devices, including my Pixel 6 Pro. After the app launches, when I tap the screen, a focus-lost event `SDL_WINDOWEVENT_FOCUS_LOST` is received in the event loop. That triggers application logic to pause gameplay. A subsequent interaction triggers `SDL_WINDOWEVENT_FOCUS_GAINED` and things proceed normally.

Bizarrely, I do not encounter this when I run debug builds via Android Studio. I only encounter this in release versions of the application. Perhaps there is a subtle lifecycle difference through the Android OS.

Here's a log trace the demonstrates the issue:

```
---------------------------- PROCESS STARTED (11701) for package com.trippinontubs ----------------------------
2023-10-08 10:46:05.520 11701-11737 SDL/APP                 com.trippinontubs                    I  window entered, window=1
2023-10-08 10:46:05.520 11701-11737 SDL/APP                 com.trippinontubs                    I  window gained keyboard focus, window=1
2023-10-08 10:46:05.520 11701-11737 SDL/APP                 com.trippinontubs                    I  window size changed, window=1
2023-10-08 10:46:05.520 11701-11737 SDL/APP                 com.trippinontubs                    I  window resized, window=1
2023-10-08 10:46:05.523 11701-11737 BLASTBufferQueue        com.trippinontubs                    E  [SurfaceView[com.trippinontubs/org.libsdl.app.SDLActivity]#1](f:0,a:0) rejecting buffer:active_size=3120x1440, requested_size=3120x1440 buffer{size=1258x2975 transform=7}
2023-10-08 10:46:06.768 11701-11701 SDL                     com.trippinontubs                    V  onWindowFocusChanged(): false
2023-10-08 10:46:06.768 11701-11701 SDL                     com.trippinontubs                    V  nativeFocusChanged()
2023-10-08 10:46:06.772 11701-11737 SDL/APP                 com.trippinontubs                    I  window lost keyboard focus, window=1
2023-10-08 10:46:10.740 11701-11701 SDL                     com.trippinontubs                    V  onWindowFocusChanged(): true
2023-10-08 10:46:10.740 11701-11701 SDL                     com.trippinontubs                    V  nativeFocusChanged()
2023-10-08 10:46:10.744 11701-11737 SDL/APP                 com.trippinontubs                    I  window gained keyboard focus, window=1
2023-10-08 10:46:12.846 11701-11701 SDL                     com.trippinontubs                    V  onWindowFocusChanged(): false
2023-10-08 10:46:12.846 11701-11701 SDL                     com.trippinontubs                    V  nativeFocusChanged()
2023-10-08 10:46:12.847 11701-11737 SDL/APP                 com.trippinontubs                    I  window lost keyboard focus, window=1
2023-10-08 10:46:13.981 11701-11701 SDL                     com.trippinontubs                    V  onPause()
2023-10-08 10:46:13.990 11701-11701 SDL                     com.trippinontubs                    V  surfaceDestroyed()
2023-10-08 10:46:13.990 11701-11701 SDL                     com.trippinontubs                    V  nativePause()
---------------------------- PROCESS ENDED (11701) for package com.trippinontubs ----------------------------
```

- `2023-10-08 10:46:05.520`: app launch
- `2023-10-08 10:46:06.768`: finger touch triggered focus loss
- `2023-10-08 10:46:10.740`: finger touch triggered focus gain
- `2023-10-08 10:46:12.846`: closed app

Do you know what might be happening?

Just as important, can you think of a way to debug this more effectively?

**Comment by AntTheAlchemist** [comment]
Just fyi, I've been launching apps on Android with SDL for a few years now and haven't experienced stray `SDL_WINDOWEVENT_FOCUS_LOST` events. I'm using the latest SDL3 at the moment and can't reproduce it, even from Play Store apps. 

What hints are you using? Anything like the ones that convert touch events to mouse or visa versa? Must be something unique to your build. Can you reproduce it be side-loading the release APK?

**Comment by ebarlas** [comment]
Thanks for the reply, @AntTheAlchemist. It's so odd. I've been testing this app on Android for years and I have never, not even once, noticed this issue until release via Google Play. Needless to say it's quite jarring to see pause-resume behavior simply by tapping the screen.

I haven't side-loaded the AAB. I was hoping Android Studio would help with deploying the release to my device, but I didn't see that option after looking around. I'll spend more time on this next.

I'm using the window flags `SDL_WINDOW_SHOWN`, `SDL_WINDOW_FULLSCREEN` and the renderer flags `SDL_RENDERER_ACCELERATED`, `SDL_RENDERER_PRESENTVSYNC`.

**Comment by AntTheAlchemist** [comment]
Debugging via Play Store isn't viable, so in Android Studio you can build and run the release version. Look for the "Build Variants" tab (usually on the left), where you'll see a "Debug/Release" dropdown. You can change the ABI here as well.

Are you using any kind of overlay? Like ads? What's common between the devices that trigger it and not?

This is odd (aren't most bugs?). What could generate a focus loss after a tap. Why would this only work on the release version and not debug version.

**Comment by ebarlas** [comment]
Thanks, @AntTheAlchemist. I deployed a release build using the build variants configuration. I cannot reproduce it, unfortunately. One thing that is different with the released version from Google Play is the game pad overlay inserted by Android OS seen here. I haven't had a moment to look into it, but now I suspect it's related.

![Screenshot_20231008-161957](https://github.com/libsdl-org/SDL/assets/221014/2157bae9-98f6-48af-8f5d-91c9fcef9260)

**Comment by ebarlas** [comment]
Sure enough!

I just disabled "Game Dashboard" on my Pixel 6 Pro and the issue disappears!

![Uploading Screenshot_20231008-163010.png…]()

**Comment by ebarlas** [comment]
![Screenshot_20231008-163010](https://github.com/libsdl-org/SDL/assets/221014/6e468235-e701-4170-96fb-f1cf2d98b3b7)

**Comment by AntTheAlchemist** [comment]
The Java-side SDL code does some magic and tweaking to get a drawable layer. It might be what's upsetting an overlay? I just played with the "Gaming Mode" option on my Modded Nokia 8, which overlays a little menu. It's not causing a problem until I double-tap to bring up the menu, then I get a focus lost. 

Can you try my "Game Controller Tester" app on Google Play? It's built with SDL3, but should still show up if this is a wider issue or not. My app has an option in settings (enable expert mode) that will display SDL "System Events". You'll see focus gained / lost events in real time. See if this game pad overlay is producing the focus lost event in my app. Keep in mind that most event names have been renamed in SDL3, but you'll figure out what they compared in SDL2.

Make sure you halt all rendering if you get a `SDL_APP_WILLENTERBACKGROUND` event. Resume when you get a `SDL_APP_DIDENTERFOREGROUND`. Probably won't make a difference.

**Comment by ebarlas** [comment]
I just installed the app and re-enabled Game Dashboard. However, I can't get the overlay to pop.

Is Game Controller Tester classified as a game in Google Play? What is the target SDK? I suspect something about the app disqualifies it.

**Comment by AntTheAlchemist** [comment]
It's not classed as a game (no `android:isGame="true"` in the manifest). There's no way to force the overlay? In that case, try a game I uploaded which was build with SDL2 from July 2022. PowBall Renaissance (Demo). See if you can break the pause game logic; this should give us some insights.

This game uses the following hints which might affect the bug:
```
	SDL_SetHint(SDL_HINT_TOUCH_MOUSE_EVENTS, "0");
	SDL_SetHint(SDL_HINT_MOUSE_TOUCH_EVENTS, "0");```

**Comment by ebarlas** [comment]
My game doesn't use the `android:isGame` attribute either. However, it is classified as a game in Google Play Console.

I'll give the other game a try shortly.

<img width="915" alt="Screenshot 2023-10-09 at 8 18 44 PM" src="https://github.com/libsdl-org/SDL/assets/221014/fbd4e463-eb2c-43c7-af0f-05666da352ef">

**Comment by AntTheAlchemist** [comment]
If you add the game tag to your manifest (which you should do anyway), you may be able to debug the issue because it may trigger the overlay of it knows it's a game outside of Play Store.

**Comment by 1bsyl** [comment]
I don't really get the current problem. I've just tried with my phone.
When there is the game overlay that some android os had, it indeeds trigger a Focus Lost / Gained. (when you open close, the menu).
Which sounds ok to me.


on very old android (< 7), SDActivity.mHasMultiWindow is false. and that would trigger a Pause state for SDL.
otherwise, you should just get the focus lost / gained events.

**Comment by ebarlas** [comment]
@1bsyl , the problem is that a focus-lost event is triggered simply by touching the screen after game launch when there has been no apparent interaction with the game dashboard overlay.

**Comment by ebarlas** [comment]
Unfortunately, that didn't work @AntTheAlchemist. Setting the attribute has no effect. It still doesn't occur during testing.

**Comment by ebarlas** [comment]
Here's a recording of the issue. I hope it clarifies things. Notice the tap gestures that trigger the pause and notice the visible gamepad overlay when I swipe up at the end.

https://github.com/libsdl-org/SDL/assets/221014/a2dfca1a-5d98-4189-96ff-1ba9fc390c9d

**Comment by 1bsyl** [comment]
maybe the tap gesture is showing the status bar, so there is a quick focus lost/gained ?

can you make the status bar on by default (modifying SDLActivity.java), removing immersive mode, to see if that helps

(also, you don't have to implement focus lost/gained to pause you app...)

**Comment by ebarlas** [comment]
@1bsyl , this issue only occurs when the Game Dashboard is enabled. It instantly goes away when disabled. And I can reproduce it every time when GD is enabled.

I can't easily test with different configurations because GD is only available when this app is distributed via Google Play.

I simply cannot reproduce when running builds from Android Studio.

**Comment by AntTheAlchemist** [comment]
@ebarlas did you try to reproduce it with my PowBall Renaissance Demo game? It's classified as a game, so should trigger GD in the same way your game does from Play Store. Then we can work out if it's a wider issue, or a problem with your project.

As a side note, you do need to pause the game when focus is lost on most platforms, but look at `SDL_HINT_ANDROID_BLOCK_ON_PAUSE`; this hint, enabled by default, should pause the app automatically. But you should also make sure that you're not doing any kind of display calls after you receive a `SDL_EVENT_WILL_ENTER_BACKGROUND`. I don't believe this affects your bug, but it's worth noting.

**Comment by ebarlas** [comment]
I have not tried your game yet but I intend to today.

That's great information about hints and events. Maybe I can handle events slightly differently and sidestep the issue that way.

**Comment by 1bsyl** [comment]
@ebarlas ok..
what is strange also, is that you're not touching the GD icon ... maybe that GD hasn't a great implementation

**Comment by ebarlas** [comment]
@AntTheAlchemist , @1bsyl it happens with the PowBall game as well! It is not just Trippin on Tubs. _Exactly_ the same pause and resume behavior.

**Comment by AntTheAlchemist** [comment]
> it happens with the PowBall game as well! It is not just Trippin on Tubs. _Exactly_ the same pause and resume behavior.

Now that is annoying! I'm not changing my pause/resume logic - because it is sound and works on all other platforms.

@ebarlas Last question (for now). Does this happen with other games? Non-SDL games?

**Comment by 1bsyl** [comment]
@ebarlas 
how can it pause the game ? 
I only see a focus lost even. which directly comes for activity state.

so you mapped your game pause logic with focus lost event ?


maybe, give a try a with removing the immersive mode. again just to make sure.

**Comment by ebarlas** [comment]
Yes, exactly. All threads and processing are paused explicitly when focus is lost.

And the application logic is likely the same in @AntTheAlchemist 's PowBall Renaissance game.

**Comment by AntTheAlchemist** [comment]
Yes, pausing on a focus lost event is the only way to stop the game continuing in the background other platforms. I'm not aware of a need to put a special `#ifdef __ANDROID__` case to change the pause logic specifically for Android, so the pause logic is sound. Without it, I've heard games running in the background on certain Android devices before now.

**Comment by 1bsyl** [comment]
Indeed the background event, is only for android/ios.
But, there is the "minimized" (maximized) event, which tells the window is not visible.
With the "focus" event, the window is still visible, and you can also have multiple window on android (and chromeos).

( I personally don´t use the focus lost/gained on android. (I looked, and yes, I have an `#ifdef __ANDROID__`) )

In this case, I agree the focus event behavior is strange, but doesn't sound a bug in SDL ...

BTW, there was also this proposition https://github.com/libsdl-org/SDL/pull/7846
to add an event when the window is occluded ...

**Comment by AntTheAlchemist** [comment]
Another caveat about only using the minimize event on Android for pause logic: this will fall over on Chromebooks, which offer a desktop type environment.

**Comment by 1bsyl** [comment]
ok, if you feel some events are not correct. feel free to submit a PR.
(maybe also to document the ifdef for Android)
but maybe, you don't want the same pause logic for desktop, and the mobile.

testing again on my phone with my app, I got the focus lost when the overlay "game turbo"  icon is pressed (that's how it's name on xiaomi). not when touching elsewhere.
Btw, I get no overlay with "Trippin on tubs" (just downloaded).

**Comment by slouken** [comment]
I'm going to chalk this up to an issue with the Game Dashboard. If you find out differently, please let us know!
